nomad/structs/acl.go

```diff
@@ -0,0 +1,147 @@
+
+	// The human friendly name of an ACL token cannot exceed 256 characters.
+	if len(a.Name) > maxTokenNameLength {
+		mErr.Errors = append(mErr.Errors, errors.New("token name too long"))
+	}
+
+	// The type of an ACL token must be set. An ACL token of type client must
+	// have associated policies, whereas a management token cannot be
+	// associated with policies.
+	switch a.Type {
+	case ACLClientToken:
+		if len(a.Policies) == 0 {
+			mErr.Errors = append(mErr.Errors, errors.New("client token missing policies"))
+		}
+	case ACLManagementToken:
+		if len(a.Policies) != 0 {
+			mErr.Errors = append(mErr.Errors, errors.New("management token cannot be associated with policies"))
+		}
+	default:
+		mErr.Errors = append(mErr.Errors, errors.New("token type must be client or management"))
+	}
+
+	// There are different validation rules depending on whether the ACL token
+	// is being created or updated.
+	switch existing {
+	case nil:
+		if a.ExpirationTTL < 0 {
+			mErr.Errors = append(mErr.Errors,
+				fmt.Errorf("token expiration TTL '%s' should not be negative", a.ExpirationTTL))
+		}
+
+		if a.ExpirationTime != nil && !a.ExpirationTime.IsZero() {
+
+			if a.CreateTime.After(*a.ExpirationTime) {
+				mErr.Errors = append(mErr.Errors, errors.New("expiration time cannot be before create time"))
+			}
+
+			// Create a time duration which details the time-til-expiry, so we can
+			// check this against the regions max and min values.
+			expiresIn := a.ExpirationTime.Sub(a.CreateTime)
+			if expiresIn > maxTTL {
+				mErr.Errors = append(mErr.Errors,
+					fmt.Errorf("expiration time cannot be more than %s in the future (was %s)",
+						maxTTL, expiresIn))
+
+			} else if expiresIn < minTTL {
+				mErr.Errors = append(mErr.Errors,
+					fmt.Errorf("expiration time cannot be less than %s in the future (was %s)",
+						minTTL, expiresIn))
+			}
+		}
+	default:
+		if existing.Global != a.Global {
+			mErr.Errors = append(mErr.Errors, errors.New("cannot toggle global mode"))
+		}
+		if existing.ExpirationTTL != a.ExpirationTTL {
+			mErr.Errors = append(mErr.Errors, errors.New("cannot update expiration TTL"))
+		}
+		if existing.ExpirationTime != a.ExpirationTime {
+			mErr.Errors = append(mErr.Errors, errors.New("cannot update expiration time"))
+		}
+	}
+
+	return mErr.ErrorOrNil()
+}
+
+// HasExpirationTime checks whether the ACL token has an expiration time value
+// set.
+func (a *ACLToken) HasExpirationTime() bool {
+	if a == nil || a.ExpirationTime == nil {
+		return false
+	}
+	return !a.ExpirationTime.IsZero()
+}
+
+// IsExpired compares the ACLToken.ExpirationTime against the passed t to
+// identify whether the token is considered expired. The function can be called
+// without checking whether the ACL token has an expiry time.
+func (a *ACLToken) IsExpired(t time.Time) bool {
+
+	// This is the fastest exit point and means we don't need to perform any
+	// expiration checks.
+	if t.IsZero() || !a.HasExpirationTime() {
```

Comment:
Is there a use case for checking `IsExpired` against a 0 time, and expecting the result to be false? Maybe it'd be safer to assume the token is expired in that case?